mmaction/models/heads/ssn_head.py

```diff
@@ -0,0 +1,417 @@
+        x1 = x0 + self.num_segments_list[1]
+        num_segments = x1 + self.num_segments_list[2]
+
+        feat_dim = x.size(1)
+        x = x.view(-1, num_segments, feat_dim)
+        num_samples = x.size(0)
+
+        scale_factors = scale_factors.view(-1, 2)
+
+        def extract_stage_feature(stage_feat, stage_parts, num_multipliers,
+                                  scale_factors):
+            """Extract stage feature based on structured temporal pyramid
+            pooling.
+
+            Args:
+                stage_feat (torch.Tensor): Stage features to be STPP.
+                stage_parts (tuple): Config of STPP.
+                num_multipliers (int): Total number of parts in the stage.
+                scale_factors (list): Ratios of the effective sampling lengths
+                    to augmented lengths.
+
+            Returns:
+                torch.Tensor: Features of the stage.
+            """
+            stage_stpp_feat = []
+            stage_len = stage_feat.size(1)
+            for stage_part in stage_parts:
+                ticks = torch.arange(0, stage_len + 1e-5,
+                                     stage_len / stage_part).int()
+                for i in range(stage_part):
+                    part_feat = stage_feat[:, ticks[i]:ticks[i + 1], :].mean(
+                        dim=1) / num_multipliers
+                    if scale_factors is not None:
+                        part_feat = (
+                            part_feat * scale_factors.view(num_samples, 1))
+                    stage_stpp_feat.append(part_feat)
+            return stage_stpp_feat
+
+        stage_stpp_feats = []
+        stage_stpp_feats.extend(
+            extract_stage_feature(x[:, :x0, :], self.stpp_stages[0],
+                                  self.multiplier_list[0], scale_factors[:,
+                                                                         0]))
+        stage_stpp_feats.extend(
+            extract_stage_feature(x[:, x0:x1, :], self.stpp_stages[1],
+                                  self.multiplier_list[1], None))
+        stage_stpp_feats.extend(
+            extract_stage_feature(x[:, x1:, :], self.stpp_stages[2],
+                                  self.multiplier_list[2], scale_factors[:,
+                                                                         1]))
+        stpp_feat = torch.cat(stage_stpp_feats, dim=1)
+
+        course_feat = x[:, x0:x1, :].mean(dim=1)
+        return course_feat, stpp_feat
+
+
+class STPPTest(nn.Module):
+    """Structured temporal pyramid pooling for SSN at testing.
+
+    Args:
+        num_classes (int): Number of classes to be classified.
+        with_regression (bool): Whether to perform regression or not.
+            Default: True.
+        stpp_stage (tuple): Config of structured temporal pyramid pooling.
+            Default: (1, (1, 2), 1).
+    """
+
+    def __init__(self,
+                 num_classes,
+                 with_regression=True,
```

Comment:
`with_regression` will cause an ambiguity with `with` handle like `with torch.no_grad():`. We can use `has_regression`, `use_regression`